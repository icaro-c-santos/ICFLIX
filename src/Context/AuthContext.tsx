import { createContext, useEffect, useState } from "react";

export type TAuthData = {
  login?: string;
  token?: string;
  refreshToken?: string;
  name?: string;
  avatarUrl?: string;
  isLoggedIn?: boolean;
};

export type TAuthContext = {
  userLogged: TAuthData;
  setUserLogged: (userLogged: TAuthData) => void;
};

export const AuthContext = createContext<TAuthContext>({} as TAuthContext);

export const AuthProvider = (props: { children: any }) => {
  const [userLogged, setUserLogged] = useState<TAuthData>({});

  useEffect(() => {
    const data = localStorage.getItem("userLogged");
    if (data) {
      const user: TAuthData = JSON.parse(data);
      user.isLoggedIn && setUserLogged(user);
    } else {
      setUserLogged({});
    }
  }, []);

  return (
    <AuthContext.Provider value={{ userLogged, setUserLogged } as TAuthContext}>
      {props.children}
    </AuthContext.Provider>
  );
};
