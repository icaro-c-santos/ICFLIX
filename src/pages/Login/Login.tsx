import { Box, Button, TextField } from "@mui/material";
import { useContext, useEffect } from "react";
import { AuthContext } from "../../Context/AuthContext";
import { useState } from "react";
import clientAuth from "../../Client/User";
import { useNavigate } from "react-router";
import AlertDialog from "../../Components/ModalError/ModalError";
import { useGoogleLogin, useGoogleOneTapLogin } from "@react-oauth/google";
import ButtonBase from "@mui/material/ButtonBase";
import IconGoogle from "../../imgs/icon-google.svg";
export const Login = () => {
  const { userLogged, setUserLogged } = useContext(AuthContext);
  const [isValidLogin, setIsValidLogin] = useState(false);
  const [login, setLogin] = useState("");
  const [password, setPassowrd] = useState("");
  const [errorTextLogin, setErrorTextLogin] = useState("");
  const [messageAlert, setMessageAlert] = useState("");
  const [openModal, setOpenModal] = useState(false);
  const navigate = useNavigate();
  const [show, setShow] = useState(true);
  const loginGoogle = useGoogleLogin({
    scope: "profile",
    onSuccess: (tokenResponse) => {
      fetch(
        `https://www.googleapis.com/oauth2/v1/userinfo?access_token=` +
          tokenResponse.access_token
      )
        .then((data) => data.json())
        .then((data) => {
          const user = {
            ...userLogged,
            avatarUrl: data.picture,
            name: data.given_name,
            isLoggedIn: true,
          };
          setUserLogged(user);
          localStorage.setItem("userLogged", JSON.stringify(user));
          navigate("/");
        });
    },
  });

  useEffect(() => {
    login.length <= 0 && setIsValidLogin(true);
  }, []);

  userLogged.isLoggedIn && navigate("/");

  const validLogin = (value: string) => {
    if (value.length <= 0) {
      setIsValidLogin(false);
      setErrorTextLogin("LOGIN NÃO PODE FICAR EM BRANCO!");
    } else {
      setIsValidLogin(true);
      setErrorTextLogin("");
    }

    return true;
  };

  const handlerButtonLogin = async () => {
    try {
      if (login.length <= 0) {
        throw new Error("DIGITE SEU LOGIN!");
      }
      if (password.length <= 0) {
        throw new Error("DIGITE SUA SENHA!");
      }
      const user = await clientAuth.loginUser({
        login: login,
        password: password,
      });
      user.isLoggedIn = true;
      setUserLogged(user);
      localStorage.setItem("userLogged", JSON.stringify(user));
      navigate("/");
    } catch (error) {
      setOpenModal(true);
      setMessageAlert(error.message);
    }
  };

  const handlerLogin = (
    event: React.ChangeEvent<HTMLInputElement | HTMLTextAreaElement>
  ) => {
    setLogin(event.target.value);
    validLogin(event.target.value);
  };

  const handlerPassword = (
    event: React.ChangeEvent<HTMLInputElement | HTMLTextAreaElement>
  ) => {
    setPassowrd(event.target.value);
    validLogin(login);
  };

  return (
    <Box minHeight={"534.5px"}>
      <AlertDialog
        message={messageAlert}
        openModal={openModal}
        setOpenModal={setOpenModal}
      />
      <Box
        sx={{
          display: "flex",
          flexDirection: "column",
          maxWidth: "400px",
          margin: "50px auto",
          justifyContent: "center",
          bgcolor: "#e6e6e6",
          padding: "40px",
          borderRadius: "25px",
          gap: "20px",
          boxShadow: "0px 2px 13px 0px #0000008f",
        }}
      >
        <Box
          sx={{
            gap: "20px",
            justifyContent: "center",
            display: "flex",
            flexDirection: "column",
          }}
        >
          <TextField
            placeholder="Login"
            onChange={handlerLogin}
            value={login}
            error={!isValidLogin}
            helperText={errorTextLogin}
          ></TextField>
          <TextField
            placeholder="Senha"
            onChange={handlerPassword}
            value={password}
            type={"password"}
          ></TextField>
        </Box>
        <Box sx={{ display: "flex", justifyContent: "space-between" }}>
          <Button
            onClick={() => {
              navigate("/register");
            }}
            variant="outlined"
            sx={{ padding: "10px", borderRadius: "10px", width: "180px" }}
          >
            Registrar
          </Button>
          <Button
            variant="outlined"
            sx={{ padding: "10px", borderRadius: "10px", width: "180px" }}
            onClick={handlerButtonLogin}
          >
            Entrar
          </Button>
        </Box>
        <Button
          onClick={() => {
            loginGoogle();
          }}
          variant="outlined"
          sx={{
            padding: "10px",
            minWidth: "300px",
            margin: "auto",
            borderRadius: "10px",
            width: "180px",
            bgcolor: "white",
            color: "blue",
            fontWeight:600
          }}
        >
          {" "}
          Faça login com o Google{" "}
          {
            <img
              style={{ marginLeft: "15px" }}
              width={"25px"}
              src={IconGoogle}
            ></img>
          }
        </Button>
      </Box>
      <Box sx={{ margin: "auto", textAlign: "center" }}>
        {
          "ATENÇÃO PARA TESTE USE O SEGUINTE USUÁRIO -> Usuario: user ,Senha: user"
        }
      </Box>
    </Box>
  );
};
