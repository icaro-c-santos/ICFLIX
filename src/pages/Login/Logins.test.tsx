import { render, screen } from "@testing-library/react";
import { Route, Routes, useNavigate } from "react-router";
import { AuthProvider } from "../../Context/AuthContext";
import {
  clearMockAuthenticUser,
  mockAuthenticUser,
} from "../../tests/mocks/localStorageMock";
import { Login } from "./Login";
import renderer from "react-test-renderer";
import { Link } from "react-router-dom";

const mockNavigate = jest.fn();
jest.mock("react-router", () => ({
  ...jest.requireActual("react-router"),
  useNavigate: () => mockNavigate,
}));

describe("<Login>", () => {
  it("should call navigate to home if logged", () => {
    mockAuthenticUser();
    render(
      <AuthProvider>
        <Login></Login>
      </AuthProvider>
    );
    expect(mockNavigate).toHaveBeenCalledWith("/");

    clearMockAuthenticUser();
  });

  it("should call navigate to home if logged", () => {
    clearMockAuthenticUser();
    const { baseElement } = render(
      <AuthProvider>
        <Login></Login>
      </AuthProvider>
    );

    screen.logTestingPlaygroundURL();
    expect(mockNavigate).not.toHaveBeenCalled();
  });

  it("Matches DOM Snapshot", () => {
    const domTree = renderer
      .create(
        <AuthProvider>
          <Login></Login>
        </AuthProvider>
      )
      .toJSON();
    expect(domTree).toMatchSnapshot();
  });
});
