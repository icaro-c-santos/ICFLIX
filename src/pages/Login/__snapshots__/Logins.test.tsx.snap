// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Login> Matches DOM Snapshot 1`] = `
Array [
  <div />,
  <div
    className="MuiBox-root css-hhdv6n"
  >
    <div
      className="MuiBox-root css-1yrfh49"
    >
      <div
        className="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
      >
        <div
          className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary Mui-error MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
          onClick={[Function]}
        >
          <input
            aria-invalid={true}
            autoFocus={false}
            className="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
            disabled={false}
            id=":r1:"
            onAnimationStart={[Function]}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            placeholder="Login"
            required={false}
            type="text"
            value=""
          />
          <fieldset
            aria-hidden={true}
            className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              className="css-ihdtdm"
            >
              <span
                className="notranslate"
              >
                ​
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
      <div
        className="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
      >
        <div
          className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
          onClick={[Function]}
        >
          <input
            aria-invalid={false}
            autoFocus={false}
            className="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
            disabled={false}
            id=":r2:"
            onAnimationStart={[Function]}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            placeholder="Senha"
            required={false}
            type="password"
            value=""
          />
          <fieldset
            aria-hidden={true}
            className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              className="css-ihdtdm"
            >
              <span
                className="notranslate"
              >
                ​
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
    </div>
    <div
      className="MuiBox-root css-gg4vpm"
    >
      <button
        className="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium css-p00234-MuiButtonBase-root-MuiButton-root"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onContextMenu={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        Registrar
      </button>
      <button
        className="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium css-p00234-MuiButtonBase-root-MuiButton-root"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onContextMenu={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        Entrar
      </button>
    </div>
  </div>,
]
`;
